File: JINSEOP/프로그래머스_WEEK_16/택배 상자.cpp
#include <string>
#include <vector>
#include <stack>
using namespace std;

stack<int> support_container;
int count_loadable_box(vector<int> order) {
    int curr_box = 1, iter = 0, cnt = 0;
    int size = order.size();

    while (true) {
        if (iter >= size) break;
        if (curr_box > size + 1) break;

        if (order[iter] == curr_box) {
            iter++;
            curr_box++;
            cnt++;
        }
        else {
            if (!support_container.empty()) {
                if (support_container.top() == order[iter]) {
                    iter++;
                    cnt++;
                    support_container.pop();
                    continue;
                }
            }

            support_container.push(curr_box);
            curr_box++;
        }
    }
    
    return cnt;
}

int solution(vector<int> order) {
    int answer = 0;
    
    answer = count_loadable_box(order);
    return answer;
}

File: DONGUK/프로그래머스_WEEK_16/택배 상자.py
from collections import deque

def solution(order):
    answer = 0
    now = 1
    max_val = max(order)
    
    truck = []
    box = []
    
    i = 0
    while True:
        if now > max_val:
            break
            
        if now == order[i]:
            truck.append(order[i])
            now += 1
            i += 1
        
        else:
            if box:
                while box and box[-1] == order[i]:
                    box.pop()
                    truck.append(order[i])
                    i += 1
                    
                
            box.append(now)        
            now += 1
    
    while box:
        top = box.pop()
        
        if top == order[i]:
            truck.append(top)
            i += 1
        
        else:
            break
            
    answer = len(truck)
    return answer